lib/src/deploy.rs

```diff
@@ -340,6 +347,47 @@ pub(crate) fn switch_origin_inplace(root: &Dir, imgref: &ImageReference) -> Resu
     Ok(newest_deployment)
 }
 
+pub fn get_kargs(repo: &ostree::Repo, fetched: &ImageState) -> Result<Vec<String>> {
+    let cancellable = gio::Cancellable::NONE;
+    let (fetched_tree, _) = repo.read_commit(fetched.ostree_commit.as_str(), cancellable)?;
+    let fetched_tree = fetched_tree.resolve_relative_path("/usr/lib/bootc/kargs.d");
+    let fetched_tree = fetched_tree.downcast::<ostree::RepoFile>().expect("downcast");
+    match fetched_tree.query_exists(cancellable) {
+        true => {}
+        false => {
+            return Ok(vec![]);
+        }
+    }
+    
+    let mut kargs = vec![];
+    let queryattrs = "standard::name,standard::type";
+    let queryflags = gio::FileQueryInfoFlags::NOFOLLOW_SYMLINKS;
+    let fetched_iter = fetched_tree.enumerate_children(queryattrs, queryflags, cancellable)?;
+    while let Some(fetched_info) = fetched_iter.next_file(cancellable)? {
+        let fetched_child = fetched_iter.child(&fetched_info);
+        let fetched_child = fetched_child.downcast::<ostree::RepoFile>().expect("downcast");
+        fetched_child.ensure_resolved()?;
+        let fetched_contents_checksum = fetched_child.checksum();
+        let f = ostree::Repo::load_file(repo, fetched_contents_checksum.as_str(), cancellable)?;
+        let file_content = f.0;
+        let mut buffer = vec![];
+        let mut reader = ostree_ext::prelude::InputStreamExtManual::into_read(file_content.unwrap());
+        let _ = std::io::Read::read_to_end(&mut reader, &mut buffer);
+        let s = std::string::String::from_utf8(buffer)?;
+        kargs.push(s);
+    }
+
+    let existing_kargs = ostree::KernelArgs::new();
+    ostree::KernelArgs::append_proc_cmdline(
+        &existing_kargs,
+        cancellable,
+    )?;
+    let mut existing_kargs = existing_kargs.to_strv().iter().map(|s| { s.as_str().to_string() }).collect();
+    kargs.append(&mut existing_kargs);
```

Comment:
@cgwalters I use `append_proc_cmdline()` to get the existing kargs on the system, append them to the kargs owned by the image, and remove any duplicates. However, I have a few questions about the next step:

1. When both an external and container-owned karg is specified, but has different values, which one wins?
2. When I remove a previously defined karg in the container image and then upgrade, it still keeps the previously defined karg (since it is now part of the system and will be found by `append_proc_cmdline()`). Is this the desired behaviour or do we want to ensure that the karg is removed?
3. Along the same stream as question 2, if we switch to a container image with a unique karg and then proceed to switch to another container image with another unique karg, what is the desired behaviour? Do we remove the old container karg or do we keep both?

Reply:
> When I remove a previously defined karg in the container image and then upgrade, it still keeps the previously defined karg (since it is now part of the system and will be found by append_proc_cmdline()). Is this the desired behaviour or do we want to ensure that the karg is removed?

This is exactly the problem right now and why it's not as easy as this.  That approach would create [hysteresis](https://blog.verbum.org/2020/08/22/immutable-%E2%86%92-reprovisionable-anti-hysteresis/) and it's very much the type of problem that bites people badly and image-based systems should be aiming to solve.

Any solution here is going to need to do something a bit more like what the OpenShift MCO does (and a bit like ostree's 3-way merge of `/etc`):

- Look at the previous image's `/usr/lib/bootc/kargs.d`
- Look at the new image's `/usr/lib/bootc/kargs.d`

Compute the difference between those two sets, and apply the diff to the new target.

Reply:
Alternatively of course, we could try hard to have some *other* state file (besides the bootloader ones) that define "local system kargs".  But that brings back shades of `/etc/default/grub` which is just confusing...and we have things that directly write those karg files today (including of course `rpm-ostree kargs` etc.).